src/CLI/L1AddressOrAlias.hs

```diff
@@ -0,0 +1,128 @@
+parseKindedAlias = fmap mkAlias . stripPrefix (pretty (demote @kind) <> ":")
+
+instance HasCLReader L1AddressOrAlias where
+  getReader = do
+    addrOrAlias <- Opt.str
+    case parseAddress addrOrAlias of
+      Right (MkAddress (addr :: KindedAddress kind')) -> case addr of
+        ImplicitAddress{} -> pure $ L1AOAKindSpecified $ AddressResolved addr
+        ContractAddress{} -> pure $ L1AOAKindSpecified $ AddressResolved addr
+        TxRollupAddress{} ->
+          Opt.readerError $ pretty $ nameF "Unexpected address kind" $
+            "expected contract or implicit address, but got tansaction rollup"
+      Left _
+        | Just alias <- parseKindedAlias @'AddressKindContract addrOrAlias
+        -> pure $ L1AOAKindSpecified $ AddressAlias alias
+        | Just alias <- parseKindedAlias @'AddressKindImplicit addrOrAlias
+        -> pure $ L1AOAKindSpecified $ AddressAlias alias
+        | otherwise
+        -> pure $ L1AOAKindUnspecified addrOrAlias
+  getMetavar = "CONTRACT OR IMPLICIT ADDRESS OR ALIAS"
+
+instance Buildable L1AddressOrAlias where
+  build = \case
+    L1AOAKindSpecified (AddressAlias (alias :: Alias kind))
+      -> (build (demote @kind) \\ aliasKindSanity alias) <> ":" +| alias |+ ""
+    L1AOAKindSpecified (AddressResolved address) -> build address
+    L1AOAKindUnspecified alias -> build alias
+
+-- | Exception thrown on ambiguous address alias.
+data AmbiguousAddressAlias = AmbiguousAddressAlias Text ContractAddress ImplicitAddress
+  deriving stock Show
+
+instance Exception AmbiguousAddressAlias where
+  displayException = pretty
+
+instance Buildable AmbiguousAddressAlias where
+  build (AmbiguousAddressAlias aliasText contractAddr implicitAddr) = [itu|
+    The alias '#{aliasText}' is assigned to both:
+      * a contract address: #{contractAddr}
+      * and an implicit address: #{implicitAddr}
+    Use '#{contractPrefix}:#{aliasText}' or '#{implicitPrefix}:#{aliasText}' to disambiguate.
+    |]
+
+-- | Try to resolve 'L1AddressOrAlias' to an 'L1Address'.
+resolveL1AddressMaybe
+  :: forall m. (MonadThrow m, HasTezosClient m)
+  => L1AddressOrAlias -> m (Maybe L1Address)
+resolveL1AddressMaybe = \case
+  L1AOAKindSpecified x -> fmap MkConstrainedAddress <$> resolveAddressMaybe x
+  L1AOAKindUnspecified x -> do
+    implicit <- resolveAddressMaybe $ AddressAlias $ mkAlias @'AddressKindImplicit x
+    contract <- resolveAddressMaybe $ AddressAlias $ mkAlias @'AddressKindContract x
+    whenJust implicit $ \implicit' -> whenJust contract $ \contract' ->
+      throwM $ AmbiguousAddressAlias x contract' implicit'
+    pure $ (MkConstrainedAddress <$> implicit) <|> (MkConstrainedAddress <$> contract)
+
+-- | Parser for 'L1AddressOrAlias'
+l1AddressOrAliasOption
+  :: Maybe L1AddressOrAlias
+  -> "name" :! String
+  -> "help" :! String
+  -> Opt.Parser L1AddressOrAlias
+l1AddressOrAliasOption dflt name (arg #help -> help) =
+  mkCLOptionParser dflt name $ #help :! List.unlines
```

Comment:
Using `unlines` doesn't help here, `optparse-applicative`'s `help` modifier will ignore newline characters.

```
$ tzbtc-client mint --help
...
  --to CONTRACT OR IMPLICIT ADDRESS OR ALIAS
                           Address to mint to When using an alias that is
                           assigned to both a contract and an implicit account,
                           use the prefix 'contract:' or 'implicit:' to
                           disambiguate.
```

You have to use the `helpDoc` modifier + `vsep` instead. This will let you have full control of the alignment.

Now, using `helpDoc` will disable the "automatic terminal-width word-wrap" behavior of optparse-applicative. But you can use [`paragraph`](https://hackage.haskell.org/package/optparse-applicative-0.17.0.0/docs/Options-Applicative-Help-Chunk.html#v:paragraph) to get that behavior back.

See example here: https://github.com/tqtezos/stablecoin/blob/bee301161fd2dbb2385d00d2ee19240dc31a1eac/haskell/src/Stablecoin/Client/L1AddressOrAlias.hs#L185-L201

Reply:
Okay, true, but actually I think I prefer the version without newlines (modulo the missing dot). Compare:

```
  --from CONTRACT OR IMPLICIT ADDRESS OR ALIAS
                           Address to transfer from
                           
                           When using an alias that is assigned to both a
                           contract and an implicit account, use the prefix
                           'contract:' or 'implicit:' to disambiguate.
  --to CONTRACT OR IMPLICIT ADDRESS OR ALIAS
                           Address to transfer to
                           
                           When using an alias that is assigned to both a
                           contract and an implicit account, use the prefix
                           'contract:' or 'implicit:' to disambiguate.
```
Passable, but the spacing is confusing.

```
  --from CONTRACT OR IMPLICIT ADDRESS OR ALIAS
                           Address to transfer from
                           When using an alias that is assigned to both a
                           contract and an implicit account, use the prefix
                           'contract:' or 'implicit:' to disambiguate.
  --to CONTRACT OR IMPLICIT ADDRESS OR ALIAS
                           Address to transfer to
                           When using an alias that is assigned to both a
                           contract and an implicit account, use the prefix
                           'contract:' or 'implicit:' to disambiguate.
```
Actually pretty bad without periods: this doesn't look like separate paragraphs, this looks like a softbreak gone bad.
```
  --from CONTRACT OR IMPLICIT ADDRESS OR ALIAS
                           Address to transfer from. When using an alias that is
                           assigned to both a contract and an implicit account,
                           use the prefix 'contract:' or 'implicit:' to
                           disambiguate.
  --to CONTRACT OR IMPLICIT ADDRESS OR ALIAS
                           Address to transfer to. When using an alias that is
                           assigned to both a contract and an implicit account,
                           use the prefix 'contract:' or 'implicit:' to
                           disambiguate.
```
Perfection :laughing: but seriously, this looks better to me.

Reply:
Ah yes, the period is definitely needed here, the paragraph is optional ^^ Yup, the last example looks great!